lib/segment/src/segment.rs

```diff
@@ -376,6 +378,53 @@ impl Segment {
             })
             .collect()
     }
+
+    pub fn filtered_read_by_index(
+        &self,
+        offset: Option<PointIdType>,
+        limit: Option<usize>,
+        condition: &Filter,
+    ) -> Vec<PointIdType> {
+        let payload_index = self.payload_index.borrow();
+        let id_tracker = self.id_tracker.borrow();
+
+        let ids_iterator = payload_index
+            .query_points(condition)
+            .filter_map(|internal_id| {
+                let external_id = id_tracker.external_id(internal_id);
+                match external_id {
+                    Some(external_id) => match offset {
+                        Some(offset) if external_id < offset => None,
+                        _ => Some(external_id),
+                    },
+                    None => None,
+                }
+            });
+
+        let mut page = match limit {
+            Some(limit) => peek_top_smallest_iterable(ids_iterator, limit),
+            None => ids_iterator.collect(),
+        };
+        page.sort_unstable();
+        page
+    }
+
+    pub fn filtered_read_by_id_stream(
+        &self,
+        offset: Option<PointIdType>,
+        limit: Option<usize>,
+        condition: &Filter,
+    ) -> Vec<PointIdType> {
+        let payload_index = self.payload_index.borrow();
+        let filter_context = payload_index.filter_context(condition);
+        self.id_tracker
+            .borrow()
+            .iter_from(offset)
+            .filter(move |(_, internal_id)| filter_context.check(*internal_id))
+            .map(|x| x.0)
```

Comment:
Maybe also destructure the tuple into `|(element1, _)|` for readability. The name `element1` is just an abstract example here

Reply:
:+1: